dev/devicelab/lib/tasks/perf_tests.dart

```diff
@@ -420,123 +389,162 @@ class CompileTest {
+  MemoryTest(this.project, this.test, this.package);
+
+  final String project;
+  final String test;
+  final String package;
+
+  Future<void> get ready => _ready?.future;
+  Completer<void> _ready;
+  String _nextMessage;
+  void prepareForNextMessage(String message) {
+    _nextMessage = message;
+    _ready = new Completer<void>();
+  }
 
-  final String testDirectory;
-  final String packageName;
+  int get iterationCount => 15;
 
-  /// Path to a flutter driver script that will run after starting the app.
-  ///
-  /// If not specified, then the test will start the app, gather statistics, and then exit.
-  final String testTarget;
+  Device get device => _device;
+  Device _device;
 
   Future<TaskResult> run() {
-    return inDirectory(testDirectory, () async {
-      final Device device = await devices.workingDevice;
+    return inDirectory(project, () async {
+      // This test currently only works on Android, because device.logcat,
+      // device.getMemoryStats, etc, aren't implemented for iOS.
+
+      _device = await devices.workingDevice;
       await device.unlock();
-      final String deviceId = device.deviceId;
       await flutter('packages', options: <String>['get']);
 
       if (deviceOperatingSystem == DeviceOperatingSystem.ios)
-        await prepareProvisioningCertificates(testDirectory);
-
-      final List<String> runOptions = <String>[
-        '-v',
-        '--profile',
-        '--trace-startup', // wait for the first frame to render
-        '-d',
-        deviceId,
-        '--observatory-port',
-        '0',
-      ];
-      if (testTarget != null)
-        runOptions.addAll(<String>['-t', testTarget]);
-      final String output = await evalFlutter('run', options: runOptions);
-      final int observatoryPort = parseServicePort(output, prefix: 'Successfully connected to service protocol: ', multiLine: true);
-      if (observatoryPort == null)
-        throw new Exception('Could not find observatory port in "flutter run" output.');
-
-      final Map<String, dynamic> startData = await device.getMemoryStats(packageName);
-
-      final Map<String, dynamic> data = <String, dynamic>{
-         'start_total_kb': startData['total_kb'],
-      };
-
-      if (testTarget != null) {
-        await flutter('drive', options: <String>[
-          '-v',
-          '-t',
-          testTarget,
-          '-d',
-          deviceId,
-          '--use-existing-app=http://localhost:$observatoryPort',
-        ]);
-
-        final Map<String, dynamic> endData = await device.getMemoryStats(packageName);
-        data['end_total_kb'] = endData['total_kb'];
-        data['diff_total_kb'] = endData['total_kb'] - startData['total_kb'];
+        await prepareProvisioningCertificates(project);
+
+      final StreamSubscription<String> adb = device.logcat.listen(
+        (String data) {
+          if (data.contains('==== MEMORY BENCHMARK ==== $_nextMessage ===='))
```

Comment:
nit: extract this pattern to a common method

Reply:
This _is_ the extracted pattern. :-) It's the only occurrence in devicelab, I think.

Unless you mean extract it from the tests... I'm not sure what a clean way to do that would be. I don't really want to put this in `flutter_test`, since it's devicelab-specific...